src/bundle/linux/common.rs

```diff
@@ -104,62 +104,80 @@ pub fn total_dir_size(dir: &Path) -> ::Result<u64> {
         }
-        let mut decoder = PNGDecoder::new(File::open(&icon_path)?);
-        let (width, height) = decoder.dimensions()?;
+    } else {
+        let icon = image::open(&icon_path)?;
+        let (width, height) = icon.dimensions();
         let is_high_density = common::is_retina(&icon_path);
+
         if !sizes.contains(&(width, height, is_high_density)) {
             sizes.insert((width, height, is_high_density));
-            let dest_path = get_dest_path(width, height, is_high_density);
-            common::copy_file(&icon_path, &dest_path)?;
+            let dest_path = get_dest_path(width, height, is_high_density, base_dir, binary_name);
+            let encoder = PNGEncoder::new(common::create_file(&dest_path)?);
+            encoder.encode(&icon.raw_pixels(), width, height, icon.color())?;
         }
     }
-    // Fall back to non-PNG files for any missing sizes.
+
+    Ok(sizes.to_owned())
+}
+
+/// Generate the icon files and store them under the `data_dir`.
+pub fn generate_icon_files(settings: &Settings, data_dir: &PathBuf) -> ::Result<()> {
+    let base_dir = data_dir.join("usr/share/icons/hicolor");
+
+    let mut sizes: BTreeSet<(u32, u32, bool)> = BTreeSet::new();
+
+    // Prefer PNG files.
     for icon_path in settings.icon_files() {
         let icon_path = icon_path?;
         if icon_path.extension() == Some(OsStr::new("png")) {
-            continue;
-        } else if icon_path.extension() == Some(OsStr::new("icns")) {
-            let icon_family = icns::IconFamily::read(File::open(&icon_path)?)?;
-            for icon_type in icon_family.available_icons() {
-                let width = icon_type.screen_width();
-                let height = icon_type.screen_height();
-                let is_high_density = icon_type.pixel_density() > 1;
-                if !sizes.contains(&(width, height, is_high_density)) {
-                    sizes.insert((width, height, is_high_density));
-                    let dest_path = get_dest_path(width, height, is_high_density);
-                    let icon = icon_family.get_icon_with_type(icon_type)?;
-                    icon.write_png(common::create_file(&dest_path)?)?;
-                }
-            }
-        } else {
-            let icon = image::open(&icon_path)?;
-            let (width, height) = icon.dimensions();
-            let is_high_density = common::is_retina(&icon_path);
-            if !sizes.contains(&(width, height, is_high_density)) {
-                sizes.insert((width, height, is_high_density));
-                let dest_path = get_dest_path(width, height, is_high_density);
-                let encoder = PNGEncoder::new(common::create_file(&dest_path)?);
-                encoder.encode(&icon.raw_pixels(), width, height, icon.color())?;
-            }
+            let new_sizes = generate_icon_files_png(&icon_path, &base_dir, settings.binary_name(), sizes.clone()).unwrap();
+            sizes.append(&mut new_sizes.to_owned())
+        }
+        {
```

Comment:
It seems to me that an else is missing here

Reply:
@body20002 or if you can explain to me what you were trying to do here, no problem :D

Reply:
I also don't fully understand this, why we need to new scope?

Reply:
I don't remember exactly why I did it, but I think I did it as a fallback to add any missing icons
in the original implementation by @anas-elgarhy he used 2 for loops one for getting the png and the other for getting non-png images, the function prefers png images then falls back to non-png if the image do not exists in png format

I added the new scope when I refactored the function, so I can have a new variable with the same name `new_sizes` :p

There is no else statement here, I'm new to rust and this was easy to do.
Please review it thoroughly and notify me if I did something wrong.

Reply:
True, it is useful for that 😃, but the problem I see here with respect to that area, is that in the case that all the icons in the folder are png, we would be doing the whole process unnecessarily 2 times, that's why I highlighted the lack of an else 🤔

Reply:
@body20002 for the new scope, U can just create a new variable with the same name and rust will take a care with the old variable and shadowing it, I don't see that we need the old variable anymore after passing it to the `append` function + I see it already in different scope

Reply:
The issue with the else that I mentioned, so that you take into account is that the function `generate_icon_files_non_png` will always be called, no matter if the icon is a png or not, this because although it is in a "separate" scope, it does not have a conditional that limits its execution

Reply:
@SergioRibera Yeah I know, I was confused why @anas-elgarhy did it in the original implementation, but I didn't ask why.
there's no else statement only 2 for loops one for generating `png` and the other for generating `non_png`

If adding an else statement will clear confusion I'll happily do it.

Reply:
@body20002 For my part, this is the only thing I would ask you to approve the PR, thank you very much 😄